PolyEngine/Core/Src/Optional.hpp

```diff
@@ -0,0 +1,188 @@
+#pragma once
+
+#include "Defines.hpp"
+
+namespace Poly {
+	template<typename V>
+	class Optional {
+	public:
+		using ValueType = V;
+	public:
+		Optional() : initialized(false), dummy{} {}
+		Optional(const V&  value) : initialized(true), value_storage(value) {}
+		Optional(      V&& value) : initialized(true), value_storage(std::move(value)) {}
+		~Optional() { if (this->HasValue()) { this->Value().~V(); } }
+
+	public:
+		Optional(Optional&& other) {
```

Comment:
You can use `Optional` in many other situations, i.e.: for fields in configuration objects, to specify lack of value, or value to use. `Optional<T>` copy constructor and copy operator don't hurt us more than `T` copy constructor. I don't see the point of this limitation.